private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,741 @@
+	}
+	return s.jsonrpc2Conn.Notify(ctx, "textDocument/publishDiagnostics", diagParams)
+}
+
+// Create a new file entry with the given contents and metadata.
+func (s *server) createFileEntry(ctx context.Context, item protocol.TextDocumentItem, resolver moduleSetResolver) (*fileEntry, error) {
+	filename := item.URI.Filename()
+	entry := newFileEntry(&item, resolver, filename, strings.HasPrefix(filename, s.container.CacheDirPath()))
+	s.fileCache[filename] = entry
+	if err := entry.processText(ctx, s); err != nil {
+		return nil, err
+	}
+	if !entry.isRemote {
+		if err := s.fileWatcher.Add(filename); err != nil {
+			return nil, err
+		}
+	}
+	return entry, nil
+}
+
+func (s *server) refreshImage(ctx context.Context, resolver moduleSetResolver) error {
+	moduleSet, err := resolver.ModuleSet()
+	if err != nil {
+		return err
+	}
+	bucket, err := resolver.Bucket()
+	if err != nil {
+		return err
+	}
+	err = bucket.WalkFileInfos(ctx, func(fileInfo bufmodule.FileInfo) error {
+		if entry, ok := s.fileCache[fileInfo.ExternalPath()]; ok {
+			entry.bufDiags = nil
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	// TODO: diagsByFile is flawed; it maps on ExternalPath, which may or may not be meaningful.
+	// The LSP should probably instead map OS paths to modules on its own and use this to tie files
+	// to diagnostics and etc.
+	diagsByFile := make(map[string][]protocol.Diagnostic)
+	image, err := bufimage.BuildImage(ctx, tracing.NewTracer(s.container.Tracer()), bucket)
+	if err != nil {
+		var fileAnnotationSet bufanalysis.FileAnnotationSet
+		if !errors.As(err, &fileAnnotationSet) {
+			return err
+		}
+		for _, annot := range fileAnnotationSet.FileAnnotations() {
+			diagsByFile[annot.FileInfo().ExternalPath()] = append(diagsByFile[annot.FileInfo().ExternalPath()],
+				annotationToDiagnostic(annot, protocol.DiagnosticSeverityError))
+		}
+	}
+	if workspace, ok := moduleSet.(bufworkspace.Workspace); ok && image != nil {
+		for _, module := range moduleSet.Modules() {
+			if err := s.lintHandler.Check(ctx, workspace.GetLintConfigForOpaqueID(module.OpaqueID()), image); err != nil {
+				var fileAnnotationSet bufanalysis.FileAnnotationSet
+				if !errors.As(err, &fileAnnotationSet) {
+					return err
+				}
+				for _, annot := range fileAnnotationSet.FileAnnotations() {
+					diagsByFile[annot.FileInfo().ExternalPath()] = append(diagsByFile[annot.FileInfo().ExternalPath()],
+						annotationToDiagnostic(annot, protocol.DiagnosticSeverityWarning))
+				}
+			}
+		}
+	}
+	for externalPath, diags := range diagsByFile {
+		if entry, ok := s.fileCache[externalPath]; ok {
+			entry.bufDiags = diags
+			if err := s.updateDiagnostics(ctx, entry); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Creates a cache path for a module key and module file path. This cache is local to the LSP.
+// The format of the cache path is:
+// <remote>/<owner>/<repository>/<commit id>/<digest type>/<digest>/<module file path>
```

Comment:
Why are we using both the commit and the digest as part of the path? You should only need one, namely the commit. It appears you are doing so so that you can construct a `ModuleKey`, but you don't actually even use the digest from the `ModuleKey`, you just call `ModuleKey.String()`, which only uses the commit ID.

Reply:
This actually comes from the standalone version of the LSP which needed it to reconstruct the module pins. Since we don't need that I can just remove the digest from the path.